llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp

```diff
@@ -81,7 +81,8 @@ bool llvm::isAllocaPromotable(const AllocaInst *AI) {
         return false;
     } else if (const IntrinsicInst *II = dyn_cast<IntrinsicInst>(U)) {
       if (!II->isLifetimeStartOrEnd() && !II->isDroppable() &&
-          II->getIntrinsicID() != Intrinsic::fake_use)
+          II->getIntrinsicID() != Intrinsic::fake_use &&
+          !II->isLaunderOrStripInvariantGroup())
```

Comment:
I don't understand how this is supposed to work. I don't think we have logic to handle these in mem2reg? Also wouldn't we have to scan their uses here?

Reply:
I might be wrong, but we don't seem to have logic to handle the others here, do we actually need any specific logic in mem2reg? Why should we scan their uses? The change seems to be needed as the alloca is marked as promotable, as otherwise we would assert in PromoteMemoryToRegister. I don't think we want to mark the alloca as non-promotable when visiting the intrinsic and prevent mem2reg from having the chance to simplify the alloca further.

Reply:
Okay, I think this gets handled here: https://github.com/llvm/llvm-project/blob/05284a071a6620cd3c5846b5ff36918da0c2e1a6/llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp#L495-L507 I think the handling is ultimately correct, but the comment there could use an update.

Reply:
Comment updated, thanks.